File: 943-sum-of-subarray-minimums/sum-of-subarray-minimums.java
class Solution {
    public static int sumSubarrayMins(int[] arr) {
        int []left=new int[arr.length];
        int []right=new int[arr.length];
        Stack<Integer>st=new Stack<>();
        int mod=1000000007;
        for(int i=0;i<arr.length;i++){
            while(!st.isEmpty() && arr[st.peek()]>arr[i]){
                st.pop();
            }
            if(st.isEmpty()){
                left[i]=-1;
            }else{
                left[i]=st.peek();
            }
            st.push(i);
        }
        st.clear();
        for(int i=arr.length-1;i>=0;i--){
            while(!st.isEmpty() && arr[st.peek()]>=arr[i]){
                st.pop();
            }
            if(st.isEmpty()){
                right[i]=arr.length;
            }else{
                right[i]=st.peek();
            }
            st.push(i);
            
        }
        long   cnt=0;
        for(int i=0;i<arr.length;i++){
           long leftCount = i - left[i];
    long rightCount = right[i] - i;
    long product = (leftCount * rightCount) % mod;
    product = (product * arr[i]) % mod;
    cnt = (cnt + product) % mod;
            
        }
        return (int)cnt;
        
        
        // code here
    }
}

File: 943-sum-of-subarray-minimums/sum-of-subarray-minimums.cpp
class Solution {
public:
    int sumSubarrayMins(vector<int>& arr) {
        vector<int>nse(arr.size()+1);
        vector<int>psee(arr.size()+1);
        stack<int>st;
        for(int i=0;i<arr.size();i++){
            while(!st.empty() && arr[st.top()]>arr[i]){
                st.pop();
            }
            psee[i]=st.empty()?-1:st.top();
            st.push(i);


        }
        stack<int>mp;
        for(int i=arr.size()-1;i>=0;i--){
            while(!mp.empty() && arr[mp.top()]>=arr[i]){
                mp.pop();
            }
            nse[i]=mp.empty()?arr.size():mp.top();
            mp.push(i);

        }
        const int mod=(int)1e9+7;
        long long sum=0;
        for(int i=0;i<arr.size();i++){
           long long product=((nse[i]-i)*(i-psee[i]))%mod;
           product=(product*arr[i])%mod;
           sum=(sum+product)%mod;


        }
        return sum;
        
    }
};